### PR TITLE
✨ [Browser SDK][RUM-291] Allow logs when cookies are disabled

### PR DESCRIPTION
**Pull request by mquentin**
What it does
Capture d’écran 2020-02-06 à 14 23 45
When the cookies are disabled, only the RUM events are blocked. All collected log events will be sent without a session id.

Context
https://datadoghq.atlassian.net/browse/RUMF-291
https://trello.com/c/ttMmQ7vw/70-browser-logs-datadog-browser-agent-doesnt-work-in-an-environment-where-cookies-are-disabled

Client have a native iOS app that opens up a web app in UIWebView that blocks cookies. As a result the  `DD_LOGS.logger.log` does not send anything as neither the session id, neither the 'Tracked or not' mode can be read from the cookies.

Changes

- init process `isValidBrowsingContext` has now two modes for each package that allow the cookies=OFF for logs and block it for rum
- The mode "Tracked/ not Tracked" is read from the cookies (current behavior) or computed on the fly from the configurations if it is not available (added behavior)
- If the cookies are off, the session id will only be a random id reloaded in every pages. When this case happens, the logs will not pass a session_id to the backend

### CODE REVIEW

File: packages/core/src/init.ts [hunk 1]
```diff
@@ -47,25 +42,20 @@ export function commonInit(userConfiguration: UserConfiguration, buildEnv: Build
...
-      return true
+export function isValidCookiesBrowsingContext() {
+  if (!areCookiesAuthorized()) {
+    console.error('Cookies are not authorized, we will not send any data.')
+    return false
+  }
+  return true
+}
+
+export function isValidLocalFileBrowsingContext() {
```
Comment (bcaudan) [inline packages/core/src/init.ts]: valid browsing context naming made sense to regroup protocol and cookie checks, here it seems a bit overkill.

what about something like `checkCookiesAuthorized` and `checkIsNotLocalFile`?

File: packages/logs/src/logger.ts [hunk 2]
```diff
@@ -70,7 +69,7 @@ export function startLogger(errorObservable: ErrorObservable, configuration: Con
             url: window.location.href,
           },
         },
-        (areCookiesAuthorized()) ? { session_id: session.getId() } : {},
+        (session.getId()) ? { session_id: session.getId() } : {},
```
Comment (bcaudan) [inline packages/logs/src/logger.ts]: you don't need this condition, `undefined` fields are removed during serialization.

```
JSON.stringify({a: undefined})
"{}"
```

File: packages/core/src/sessionManagement.ts [hunk 3]
```diff
@@ -7,7 +7,9 @@ export const EXPIRATION_DELAY = 15 * utils.ONE_MINUTE
 
 export interface Session<T> {
   renewObservable: Observable<void>
+
   getId(): string | undefined
+
   getType(): T | undefined
 }
```
Comment (bcaudan) [inline packages/core/src/sessionManagement.ts]: do you have a rule in your IDE that add these extra spaces?

File: packages/core/src/sessionManagement.ts [hunk 4]
```diff
@@ -65,7 +68,7 @@ export function startSessionManagement<Type extends string>(
 
   return {
     getId() {
-      return sessionId.get()
+      return cookiesAuthorized ? sessionId.get() : undefined
```
Comment (bcaudan) [inline packages/core/src/sessionManagement.ts]: why adding this logic in core session management? 
this logic should  only apply for logs, right?

Comment (bcaudan) [inline packages/core/src/sessionManagement.ts]: why adding this logic in core session management?
this logic should only apply for logs, right?

File: packages/core/src/sessionManagement.ts (hunk 3 above)
Comment (mquentin) [reply packages/core/src/sessionManagement.ts]: I fixed it later: I was using the default Webstorm reformat code. Now it is using the reformat prettier one. Do you want me to revert these ?

Comment (bcaudan) [reply packages/core/src/sessionManagement.ts]: yes please 🙂

File: packages/logs/src/loggerSession.ts [hunk 5]
```diff
@@ -12,27 +12,33 @@ export enum LoggerSessionType {
...
   }
+  return LoggerSessionType.TRACKED
+}
+
+function computeSessionState(configuration: Configuration, rawSessionType?: string) {
+  const sessionType = hasValidLoggerSession(rawSessionType) ? rawSessionType : computeSessionType(configuration)
   return {
     isTracked: sessionType === LoggerSessionType.TRACKED,
-    type: sessionType,
+    type: hasValidLoggerSession(rawSessionType) ? rawSessionType : computeSessionType(configuration),
```
Comment (bcaudan) [inline packages/logs/src/loggerSession.ts]: Do we still need that change?

File: packages/logs/src/loggerSession.ts [hunk 6]
```diff
@@ -12,27 +12,33 @@ export enum LoggerSessionType {
...
 
-export function startLoggerSession(configuration: Configuration): LoggerSession {
+export function startLoggerSession(configuration: Configuration, areCookieAuthorized: boolean): LoggerSession {
+  if (!areCookieAuthorized) {
+    const isTracked = computeSessionType(configuration) === LoggerSessionType.TRACKED
+    return {
+      getId: () => undefined,
+      isTracked: () => isTracked,
+    }
+  }
```
Comment (bcaudan) [inline packages/logs/src/loggerSession.ts]: Can you add a test case for that?

File: packages/logs/src/loggerSession.ts (hunk 5 above)
Comment (mquentin) [reply packages/logs/src/loggerSession.ts]: no I removed it

File: packages/logs/src/loggerSession.ts (hunk 6 above)
Comment (mquentin) [reply packages/logs/src/loggerSession.ts]: I added a test

**Pull request merged**